### ISSUE TITLE
Suggest adding example for intended use of !!! in recode

### ISSUE THREAD

**Issue by jalsalam** [tidyverse/dplyr#3218]
Unquote-splicing as feature of recode in [0.7.0](https://github.com/tidyverse/dplyr/blob/master/NEWS.md#other). It would be helpful for the documentation to include an example of the intended way to use this feature. I suggest this:

``` r

library(tidyverse)
#> -- Attaching packages --------------------------------------------------------------------------------- tidyverse 1.2.1 --
#> v ggplot2 2.2.1     v purrr   0.2.4
#> v tibble  1.3.4     v dplyr   0.7.4
#> v tidyr   0.7.2     v stringr 1.2.0
#> v readr   1.1.1     v forcats 0.2.0
#> -- Conflicts ------------------------------------------------------------------------------------ tidyverse_conflicts() --
#> x dplyr::filter() masks stats::filter()
#> x dplyr::lag()    masks stats::lag()

x <- sample(c("a", "b", "c"), 10, replace = TRUE)
dplyr::recode(x, a = "Apple")
#>  [1] "Apple" "b"     "Apple" "b"     "Apple" "b"     "c"     "b"    
#>  [9] "Apple" "b"

fruits <- c("Apple", "Banana", "Carrot")
fruit_lookup <- set_names(fruits, tolower(str_sub(fruits, 1, 1))) %>% as.list

recode(x, !!!fruit_lookup)
#>  [1] "Apple"  "Banana" "Apple"  "Banana" "Apple"  "Banana" "Carrot"
#>  [8] "Banana" "Apple"  "Banana"
```

<details><summary>Session info</summary>

``` r
devtools::session_info()
#> Session info -------------------------------------------------------------
#>  setting  value                       
#>  version  R version 3.4.2 (2017-09-28)
#>  system   x86_64, mingw32             
#>  ui       RTerm                       
#>  language (EN)                        
#>  collate  English_United States.1252  
#>  tz       America/New_York            
#>  date     2017-11-21
#> Packages -----------------------------------------------------------------
#>  package    * version date       source        
#>  assertthat   0.2.0   2017-04-11 CRAN (R 3.4.1)
#>  backports    1.1.1   2017-09-25 CRAN (R 3.4.1)
#>  base       * 3.4.2   2017-09-28 local         
#>  bindr        0.1     2016-11-13 CRAN (R 3.4.1)
#>  bindrcpp     0.2     2017-06-17 CRAN (R 3.4.1)
#>  broom        0.4.3   2017-11-20 CRAN (R 3.4.2)
#>  cellranger   1.1.0   2016-07-27 CRAN (R 3.4.1)
#>  cli          1.0.0   2017-11-05 CRAN (R 3.4.2)
#>  colorspace   1.3-2   2016-12-14 CRAN (R 3.4.1)
#>  compiler     3.4.2   2017-09-28 local         
#>  crayon       1.3.4   2017-09-16 CRAN (R 3.4.1)
#>  datasets   * 3.4.2   2017-09-28 local         
#>  devtools     1.13.4  2017-11-09 CRAN (R 3.4.2)
#>  digest       0.6.12  2017-01-27 CRAN (R 3.4.1)
#>  dplyr      * 0.7.4   2017-09-28 CRAN (R 3.4.2)
#>  evaluate     0.10.1  2017-06-24 CRAN (R 3.4.1)
#>  forcats    * 0.2.0   2017-01-23 CRAN (R 3.4.1)
#>  foreign      0.8-69  2017-06-22 CRAN (R 3.4.2)
#>  ggplot2    * 2.2.1   2016-12-30 CRAN (R 3.4.1)
#>  glue         1.2.0   2017-10-29 CRAN (R 3.4.2)
#>  graphics   * 3.4.2   2017-09-28 local         
#>  grDevices  * 3.4.2   2017-09-28 local         
#>  grid         3.4.2   2017-09-28 local         
#>  gtable       0.2.0   2016-02-26 CRAN (R 3.4.1)
#>  haven        1.1.0   2017-07-09 CRAN (R 3.4.1)
#>  hms          0.3     2016-11-22 CRAN (R 3.4.1)
#>  htmltools    0.3.6   2017-04-28 CRAN (R 3.4.1)
#>  httr         1.3.1   2017-08-20 CRAN (R 3.4.1)
#>  jsonlite     1.5     2017-06-01 CRAN (R 3.4.1)
#>  knitr        1.17    2017-08-10 CRAN (R 3.4.1)
#>  lattice      0.20-35 2017-03-25 CRAN (R 3.4.2)
#>  lazyeval     0.2.1   2017-10-29 CRAN (R 3.4.2)
#>  lubridate    1.7.1   2017-11-03 CRAN (R 3.4.2)
#>  magrittr     1.5     2014-11-22 CRAN (R 3.4.1)
#>  memoise      1.1.0   2017-04-21 CRAN (R 3.4.1)
#>  methods    * 3.4.2   2017-09-28 local         
#>  mnormt       1.5-5   2016-10-15 CRAN (R 3.4.0)
#>  modelr       0.1.1   2017-07-24 CRAN (R 3.4.1)
#>  munsell      0.4.3   2016-02-13 CRAN (R 3.4.1)
#>  nlme         3.1-131 2017-02-06 CRAN (R 3.4.2)
#>  parallel     3.4.2   2017-09-28 local         
#>  pkgconfig    2.0.1   2017-03-21 CRAN (R 3.4.1)
#>  plyr         1.8.4   2016-06-08 CRAN (R 3.4.1)
#>  psych        1.7.8   2017-09-09 CRAN (R 3.4.1)
#>  purrr      * 0.2.4   2017-10-18 CRAN (R 3.4.2)
#>  R6           2.2.2   2017-06-17 CRAN (R 3.4.1)
#>  Rcpp         0.12.13 2017-09-28 CRAN (R 3.4.2)
#>  readr      * 1.1.1   2017-05-16 CRAN (R 3.4.1)
#>  readxl       1.0.0   2017-04-18 CRAN (R 3.4.1)
#>  reshape2     1.4.2   2016-10-22 CRAN (R 3.4.1)
#>  rlang        0.1.4   2017-11-05 CRAN (R 3.4.2)
#>  rmarkdown    1.8     2017-11-17 CRAN (R 3.4.2)
#>  rprojroot    1.2     2017-01-16 CRAN (R 3.4.1)
#>  rstudioapi   0.7     2017-09-07 CRAN (R 3.4.1)
#>  rvest        0.3.2   2016-06-17 CRAN (R 3.4.1)
#>  scales       0.5.0   2017-08-24 CRAN (R 3.4.1)
#>  stats      * 3.4.2   2017-09-28 local         
#>  stringi      1.1.6   2017-11-17 CRAN (R 3.4.2)
#>  stringr    * 1.2.0   2017-02-18 CRAN (R 3.4.1)
#>  tibble     * 1.3.4   2017-08-22 CRAN (R 3.4.1)
#>  tidyr      * 0.7.2   2017-10-16 CRAN (R 3.4.2)
#>  tidyverse  * 1.2.1   2017-11-14 CRAN (R 3.4.2)
#>  tools        3.4.2   2017-09-28 local         
#>  utils      * 3.4.2   2017-09-28 local         
#>  withr        2.1.0   2017-11-01 CRAN (R 3.4.2)
#>  xml2         1.1.1   2017-01-24 CRAN (R 3.4.1)
#>  yaml         2.1.14  2016-11-12 CRAN (R 3.4.1)
```

</details>

I think something like this is a very common use case. I think it would be useful to have an example and/or test to make sure the intended idiom is stable. I assume current behavior is the intended behaviour, but I *think* that `recode` used to work with a named character vector instead of a list, but that isn't the case with the current CRAN versions. 

``` r
suppressPackageStartupMessages(library(tidyverse))

x <- sample(c("a", "b", "c"), 10, replace = TRUE)
dplyr::recode(x, a = "Apple")
#>  [1] "c"     "Apple" "c"     "b"     "Apple" "Apple" "c"     "c"    
#>  [9] "c"     "c"

fruits <- c("Apple", "Banana", "Carrot")
fruit_lookup_chr <- set_names(fruits, tolower(str_sub(fruits, 1, 1)))

recode(x, !!!fruit_lookup_chr)
#> Error: Only lists can be spliced
```

<details><summary>Session info</summary>

``` r
devtools::session_info()
#> Session info -------------------------------------------------------------
#>  setting  value                       
#>  version  R version 3.4.2 (2017-09-28)
#>  system   x86_64, mingw32             
#>  ui       RTerm                       
#>  language (EN)                        
#>  collate  English_United States.1252  
#>  tz       America/New_York            
#>  date     2017-11-21
#> Packages -----------------------------------------------------------------
#>  package    * version date       source        
#>  assertthat   0.2.0   2017-04-11 CRAN (R 3.4.1)
#>  backports    1.1.1   2017-09-25 CRAN (R 3.4.1)
#>  base       * 3.4.2   2017-09-28 local         
#>  bindr        0.1     2016-11-13 CRAN (R 3.4.1)
#>  bindrcpp     0.2     2017-06-17 CRAN (R 3.4.1)
#>  broom        0.4.3   2017-11-20 CRAN (R 3.4.2)
#>  cellranger   1.1.0   2016-07-27 CRAN (R 3.4.1)
#>  cli          1.0.0   2017-11-05 CRAN (R 3.4.2)
#>  colorspace   1.3-2   2016-12-14 CRAN (R 3.4.1)
#>  compiler     3.4.2   2017-09-28 local         
#>  crayon       1.3.4   2017-09-16 CRAN (R 3.4.1)
#>  datasets   * 3.4.2   2017-09-28 local         
#>  devtools     1.13.4  2017-11-09 CRAN (R 3.4.2)
#>  digest       0.6.12  2017-01-27 CRAN (R 3.4.1)
#>  dplyr      * 0.7.4   2017-09-28 CRAN (R 3.4.2)
#>  evaluate     0.10.1  2017-06-24 CRAN (R 3.4.1)
#>  forcats    * 0.2.0   2017-01-23 CRAN (R 3.4.1)
#>  foreign      0.8-69  2017-06-22 CRAN (R 3.4.2)
#>  ggplot2    * 2.2.1   2016-12-30 CRAN (R 3.4.1)
#>  glue         1.2.0   2017-10-29 CRAN (R 3.4.2)
#>  graphics   * 3.4.2   2017-09-28 local         
#>  grDevices  * 3.4.2   2017-09-28 local         
#>  grid         3.4.2   2017-09-28 local         
#>  gtable       0.2.0   2016-02-26 CRAN (R 3.4.1)
#>  haven        1.1.0   2017-07-09 CRAN (R 3.4.1)
#>  hms          0.3     2016-11-22 CRAN (R 3.4.1)
#>  htmltools    0.3.6   2017-04-28 CRAN (R 3.4.1)
#>  httr         1.3.1   2017-08-20 CRAN (R 3.4.1)
#>  jsonlite     1.5     2017-06-01 CRAN (R 3.4.1)
#>  knitr        1.17    2017-08-10 CRAN (R 3.4.1)
#>  lattice      0.20-35 2017-03-25 CRAN (R 3.4.2)
#>  lazyeval     0.2.1   2017-10-29 CRAN (R 3.4.2)
#>  lubridate    1.7.1   2017-11-03 CRAN (R 3.4.2)
#>  magrittr     1.5     2014-11-22 CRAN (R 3.4.1)
#>  memoise      1.1.0   2017-04-21 CRAN (R 3.4.1)
#>  methods    * 3.4.2   2017-09-28 local         
#>  mnormt       1.5-5   2016-10-15 CRAN (R 3.4.0)
#>  modelr       0.1.1   2017-07-24 CRAN (R 3.4.1)
#>  munsell      0.4.3   2016-02-13 CRAN (R 3.4.1)
#>  nlme         3.1-131 2017-02-06 CRAN (R 3.4.2)
#>  parallel     3.4.2   2017-09-28 local         
#>  pkgconfig    2.0.1   2017-03-21 CRAN (R 3.4.1)
#>  plyr         1.8.4   2016-06-08 CRAN (R 3.4.1)
#>  psych        1.7.8   2017-09-09 CRAN (R 3.4.1)
#>  purrr      * 0.2.4   2017-10-18 CRAN (R 3.4.2)
#>  R6           2.2.2   2017-06-17 CRAN (R 3.4.1)
#>  Rcpp         0.12.13 2017-09-28 CRAN (R 3.4.2)
#>  readr      * 1.1.1   2017-05-16 CRAN (R 3.4.1)
#>  readxl       1.0.0   2017-04-18 CRAN (R 3.4.1)
#>  reshape2     1.4.2   2016-10-22 CRAN (R 3.4.1)
#>  rlang        0.1.4   2017-11-05 CRAN (R 3.4.2)
#>  rmarkdown    1.8     2017-11-17 CRAN (R 3.4.2)
#>  rprojroot    1.2     2017-01-16 CRAN (R 3.4.1)
#>  rstudioapi   0.7     2017-09-07 CRAN (R 3.4.1)
#>  rvest        0.3.2   2016-06-17 CRAN (R 3.4.1)
#>  scales       0.5.0   2017-08-24 CRAN (R 3.4.1)
#>  stats      * 3.4.2   2017-09-28 local         
#>  stringi      1.1.6   2017-11-17 CRAN (R 3.4.2)
#>  stringr    * 1.2.0   2017-02-18 CRAN (R 3.4.1)
#>  tibble     * 1.3.4   2017-08-22 CRAN (R 3.4.1)
#>  tidyr      * 0.7.2   2017-10-16 CRAN (R 3.4.2)
#>  tidyverse  * 1.2.1   2017-11-14 CRAN (R 3.4.2)
#>  tools        3.4.2   2017-09-28 local         
#>  utils      * 3.4.2   2017-09-28 local         
#>  withr        2.1.0   2017-11-01 CRAN (R 3.4.2)
#>  xml2         1.1.1   2017-01-24 CRAN (R 3.4.1)
#>  yaml         2.1.14  2016-11-12 CRAN (R 3.4.1)
```

</details>

I apologize -- I was not able to successfully build development or previous versions of dplyr or rlang, so I wasn't able to confirm if this was really a change or my memory is wrong.

**Comment by krlmlr** [comment]
Thanks. The triple bang `!!!` is available only with dplyr 0.7.0, so your last example can't work with CRAN versions earlier than that. I just checked, `recode(x, fruit_lookup_chr)` doesn't work with dplyr 0.5.0 either.

An update to the programming vignette is pending (#3201), does that include the information you're looking for?

**Comment by jameelalsalam** [comment]
I see, I did not realize that examples had to be cross-version compatible. That clarifies why `!!!` doesn't show up in the examples. Thanks.

**Comment by krlmlr** [comment]
No, the examples in current dplyr don't need to run with earlier versions. I just can't replicate using `recode()` with a named character vector.

**Comment by batpigandme** [comment]
Proposed example (basically @jalsalam's made slightly more explicit with the list naming), though we might want to hold off until #3390 is fixed to add.
``` r
suppressPackageStartupMessages(library(tidyverse))
# Named lists can be used for unquote splicing
x <- sample(c("a", "b", "c"), 10, replace = TRUE)
level_key <- list(a = "apple", b = "banana", c = "carrot")
recode(x, !!! level_key)
#>  [1] "carrot" "apple"  "carrot" "apple"  "banana" "carrot" "carrot"
#>  [8] "apple"  "banana" "apple"
```
Created on 2018-03-25 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

**Comment by krlmlr** [comment]
Fix for #3390 pending in #3459.

**Comment by jalsalam** [comment]
Thank you Kirill and Mara! I really appreciate you adding this example. As someone who learns R from the doc examples, and for whom recode was the first time I figured out tidyeval, I think it will be useful. This was my first time submitting an issue. Next time I will work up the nerve for a pull request. :)
